twib/twibd/TCPFrontend.cpp

```diff
@@ -0,0 +1,261 @@
+#ifdef _WIN32
+#include<ws2tcpip.h>
+#include<winsock2.h>
+#else
+#include<sys/socket.h>
+#include<sys/select.h>
+#include<netinet/in.h>
+#endif
+
+#include "TCPFrontend.hpp"
+
+#include<algorithm>
+
+#include<string.h>
+#include<unistd.h>
+#include<iostream>
+
+#include "Twibd.hpp"
+#include "Protocol.hpp"
+#include "config.hpp"
+
+namespace twili {
+namespace twibd {
+namespace frontend {
+
+#ifdef _WIN32
+char *GetFormattedMessage() {
+	char *s = NULL;
+	FormatMessage(FORMAT_MESSAGE_ALLOCATE_BUFFER | FORMAT_MESSAGE_FROM_SYSTEM | FORMAT_MESSAGE_IGNORE_INSERTS,
+			NULL, WSAGetLastError(),
+			MAKELANGID(LANG_NEUTRAL, SUBLANG_DEFAULT),
+			(LPTSTR)&s, 0, NULL);
+	return s;
+}
+
+#define NET_ERR_STR GetFormattedMessage()
+#else
+#define NET_ERR_STR strerror(errno)
+#endif
+
+TCPFrontend::TCPFrontend(Twibd *twibd) : twibd(twibd) {
+	fd = socket(AF_INET6, SOCK_STREAM, 0);
+	if (fd < 0) {
+		log(FATAL, "failed to create TCP socket: %s", NET_ERR_STR);
+		exit(1);
+	}
+	log(DEBUG, "created TCP socket: %d", fd);
+	int ipv6only = 0;
+	if (setsockopt(fd, IPPROTO_IPV6, IPV6_V6ONLY, (char*)&ipv6only, sizeof(ipv6only)) == -1) {
+		log(FATAL, "Failed to make ipv6 server dual stack: %s", NET_ERR_STR);
+	}
+
+	struct sockaddr_in6 addr;
+	memset(&addr, 0, sizeof(addr));
+	addr.sin6_family = AF_INET6;
+	addr.sin6_port = htons(15151);
+	addr.sin6_addr = in6addr_any;
+
+	if (bind(fd, (struct sockaddr*)&addr, sizeof(addr)) < 0) {
+		log(FATAL, "failed to bind TCP socket: %s", NET_ERR_STR);
+		close(fd);
+		exit(1);
+	}
+
+	if (listen(fd, 20) < 0) {
+		log(FATAL, "failed to listen on TCP socket: %s", NET_ERR_STR);
+		close(fd);
+		exit(1);
+	}
+
+	std::thread event_thread(&TCPFrontend::event_thread_func, this);
+	this->event_thread = std::move(event_thread);
+}
+
+TCPFrontend::~TCPFrontend() {
+	event_thread_destroy = true;
+	NotifyEventThread();
+	event_thread.join();
+	close(fd);
+}
+
+void TCPFrontend::event_thread_func() {
+	fd_set readfds;
+	fd_set writefds;
+	int max_fd = 0;
+	while (!event_thread_destroy) {
+		log(DEBUG, "tcp event thread loop");
+		FD_ZERO(&readfds);
+		FD_ZERO(&writefds);
+		max_fd = std::max(max_fd, fd);
+		FD_SET(fd, &readfds);
+
+		for (auto &c : clients) {
+			FD_SET(c->fd, &readfds);
+			max_fd = std::max(max_fd, c->fd);
+			if (c->out_buffer.size() > 0) {
+				FD_SET(c->fd, &writefds);
+			}
+		}
+
+		/*if (max_fd > FD_SETSIZE) {
```

Comment:
Why is this commented out o_O

Reply:
Because windows is stupid and sets FD_SETSIZE to a stupid small value like 64, but returns stupid big file descriptor values, like 600.